### ISSUE TITLE
[feature request] torch.mix function

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#104849]
### 🚀 The feature, motivation and pitch

I propose to introduce a convenience/ux function in core:
`mix = lambda a = 1, b = 1, c = 1, A = 1, B = 1, C = 1: a*b*c + A*B*C` where all a,b,c,A,B,C can be scalars or tensors. The most frequent cases would be:
- a, A, b, B are scalars, c and C are tensors
- a, A, b are scalars, c, B, C are tensors

This would generalize/symmetrize torch.addcmul, and could be used in optimizers to make second moment calculations more idiomatic and clear. This function could be implemented using torch.compile in core. Also an inplace version is needed (kind of generalization of torch.lerp_)

Instead of `mix` , the name could also be `linear_combination` (?) (although if B=C or other cases, it would not actually be a linear combination)

Some older context:
for addcmul/second moment cases, I propose to address it with a new function: sth like `mix3` discussed in https://github.com/pytorch/pytorch/issues/71683#issuecomment-1379058082 / https://github.com/pytorch/pytorch/issues/79352#issuecomment-1152946076

**Comment by albanD** [comment]
@vadimkantorov I'm curious if you expect the main benefit here to be performance (speed/mem) or API improvement for the end user?
If it is the first, would our best friend torch.compile be able to help?

**Comment by vadimkantorov** [comment]
For me, the main benefit is touching the memory less times / less memory writes + clear/expressive/convenient UX.

Currently this pattern (multiply the accumulator by decay + add statistics squared with also a decay) is used in Adam and other optimizers to collect EMA square-gradient statistics.

This might also be a useful generalization of lerp (where the weights might not sum up to 1)

I think it would be perfect if this can be torch.compiled but caveats are need for inplace + quite polymorphic possible patterns regarding scalar/tensor and supported combinations.

Another thing to keep in mind is that it's highly agreeable if 0 * inf can skip nans (then this maybe can also be a generalization of torch.where). Currently this is a problem with torch.lerp:
- https://github.com/pytorch/pytorch/issues/71701
and is related to:
- https://github.com/pytorch/pytorch/issues/50122

such pseudo-functions or functions can probably be very well-fused with inductor-generated code

Btw, have there been attempts to torch.compile the optimizer updates? Or generate foreach-variants semi-automatically?

**Comment by vadimkantorov** [comment]
An important checkbox for this function (even if it's somehow inductor-generated) should be that it should improve perf over existing calls in Adam.

**Comment by mlazos** [comment]
You can try compiling Adam for your use case in the nightly right now.
Just call step() in a function wrapped with `@torch.compile`, I'm interested to see if it handles your use case or if there are issues.

**Comment by vadimkantorov** [comment]
In OpenCV this exists as `cv2.addWeighted` and `cv2.scaleAdd`:
- https://docs.opencv.org/4.x/d5/dc4/tutorial_adding_images.html
- https://docs.opencv.org/4.x/d2/de8/group__core__array.html#gafafb2513349db3bcff51f54ee5592a19
- https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga9e0845db4135f55dcf20227402f00d98

**Comment by vadimkantorov** [comment]
Currently in optimizer code, various addcmul / mul are used for this (especially for computing lerps of squared gradients)

**Comment by mlazos** [comment]
Your suggestion is very similar to [foreach_map](https://github.com/pytorch/pytorch/blob/06b3265cb1a27ea3090dc1fc86a6ce17518f989c/test/inductor/test_compiled_optimizers.py#L100): it uses compile underneath to compile a lambda or function and horizontally fuse. You could actually use this today to do what you're describing, it will vertically fuse by default.

**Comment by vadimkantorov** [comment]
I guess the question becomes why it's not used by core optimizers yet... Because they want to avoid compilation overhead? and same question for non-foreach code paths

I think in the code everywhere still is combination of manual `addcmul_` and `lerp_`...

In the code above it uses `def update_exp_avg_sq(exp_avg_sq, grad, beta2): return exp_avg_sq.mul(beta2).addcmul(grad, grad, value=1 - beta2)` - under compile it probably is fine...

But yes, this proposal for torch.mix is also for more clear UX...

**Comment by mlazos** [comment]
oh we could decompose that if you want, I thought I'd at least use the pre-fused operators as well for brevity. I think compile time could be a concern for these ops, especially on larger models with hundreds of buffers.

**Comment by vadimkantorov** [comment]
I also wonder if it's possible to have some sort of torch.compile guards - to assert (or fail) that this function will compile to at most one Triton kernel

I guess no need to decompose in the test... I just meant to point out that this is the pattern also found elsewhere in the optimizers code, so still the eager fused operators are somehow the least common denominator - at least for code in core... (I understand the overhead reasons etc, but it's a bit of a pity)
